### PR TITLE
ECS - Run Task (Fargate)

### PR DESCRIPTION
**Pull request by grejdi-mbta**

### CODE REVIEW

**Review by paulswartz — COMMENTED** [review]
🍰 looks fine to me; is there a workflow which uses this?

**Comment by grejdi-mbta** [comment]
@paulswartz The data platform repo will use it to run the task for deploying its db migrations (see [this](https://github.com/mbta/data_platform/blob/main/.github/workflows/deploy.yml)). Still working out some more details.

**Comment by grejdi-mbta** [comment]
@paulswartz small updates here to make things a little easier with less variables.

File: run-ecs-task/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+### Inputs
+
+* **aws_access_key_id**: (Required) Will be passed to GitHub Actions environment as the AWS_ACCESS_KEY_ID environment variable to be used in AWS CLI commands.
+* **aws_secret_access_key**: (Required) Will be passed to GitHub Actions environment as the AWS_SECRET_ACCESS_KEY environment variable to be used in AWS CLI commands.
+* **aws_default_region**: (Required) Will be passed to GitHub Actions environment as the AWS_DEFAULT_REGION environment variable to be used in AWS CLI commands (from `aws configure`).
+* **ecs_cluster**: (Required) Cluster ARN to be used for running the task (ex. `arn:aws:ecs:{aws-region}:{aws-account-id}:cluster/{cluster-name}`). 
+* **ecs_task_definition**: (Required) Task Definition ARN to base the task on (ex. `arn:aws:ecs:{aws-region}:{aws-account-id}:task-definition/{task-definition-name}`).
+* **ecs_task_network_configuration**: (Required) Network settings that the task will be run with. See [this](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-configure-network.html) for more information.
```
Comment (paulswartz) [inline run-ecs-task/README.md]: question: should this be required? it's required for tasks which use the `awsvpc` networking stack, but isn't supported for any other modes.

File: run-ecs-task/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+* **ecs_task_overrides**: (Required) Override settings that the task will be run with. See [this](https://docs.aws.amazon.com/cli/latest/reference/ecs/run-task.html) for more information.
```
Comment (paulswartz) [inline run-ecs-task/README.md]: question: does this need to be required? If you wanted to keep the script simpler, you could default it to `[]`

File: run-ecs-task/action.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+# run task from task definition
+task_run=$(aws ecs run-task --task-definition "${ECS_TASK_DEFINITION}" --cluster "${ECS_CLUSTER}" --launch-type FARGATE --network-configuration "${ECS_TASK_NETWORK_CONFIGURATION}" --overrides "${ECS_TASK_OVERRIDES}")
+# get unique task arn
+task_arn=$(echo $task_run | jq '.tasks[0].taskArn')
```
Comment (thecristen) [inline run-ecs-task/action.sh]:
```suggestion
task_arn=$(echo "$task_run" | jq '.tasks[0].taskArn')
```

File: run-ecs-task/action.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+# keep track of if finished, or timed out
+finished=false
+timed_out=false
+time=0
+# continously check for completion until either finished or timed out
+while [[ "${finished}" == "false" && "${timed_out}" == "false" ]] ; do
+
+  # get the latest state of the task
+  task_description=$(aws ecs describe-tasks --cluster "${ECS_CLUSTER}" --tasks [$task_arn])
```
Comment (thecristen) [inline run-ecs-task/action.sh]:
```suggestion
  task_description=$(aws ecs describe-tasks --cluster "${ECS_CLUSTER}" --tasks ["$task_arn"])
```

File: run-ecs-task/action.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  # get the latest status
+  task_status=$(echo $task_description | jq '.tasks[0].lastStatus')
```
Comment (thecristen) [inline run-ecs-task/action.sh]:
```suggestion
  task_status=$(echo "$task_description" | jq '.tasks[0].lastStatus')
```

File: run-ecs-task/action.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    ((time=$time+5))
+  fi
+
+done
+
+# when timed out, exit with an error code and the last task description
+if [[ "${timed_out}" == "true" ]] ; then
+  echo 'Run Task failed!'
+  echo '----------------------------------------'
+  echo $task_description
```
Comment (thecristen) [inline run-ecs-task/action.sh]:
```suggestion
  echo "$task_description"
```

**Review by thecristen — CHANGES_REQUESTED** [review]
Some tiny formatting suggestions via [Shellcheck](https://www.shellcheck.net), and I echo Paul's questions about the required variables. Otherwise lgtm

**Comment by grejdi-mbta** [comment]
@paulswartz @thecristen Thanks both. I removed the 'required' as per your suggestion.

File: run-ecs-task/action.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+# run task from task definition
+task_run=$(aws ecs run-task --task-definition "${ECS_TASK_DEFINITION}" --cluster "${ECS_CLUSTER}" --launch-type FARGATE --network-configuration "${ECS_TASK_NETWORK_CONFIGURATION}" --overrides "${ECS_TASK_OVERRIDES}")
```
Comment (thecristen) [inline run-ecs-task/action.sh]: Will this still work as expected if the inputs populating `ECS_TASK_NETWORK_CONFIGURATION` and `ECS_TASK_OVERRIDES` are undefined?

Comment (grejdi-mbta) [reply run-ecs-task/action.sh]: Will check. It was working for my use case, but it's a different setup from MBTA's infrastructure.

**Comment by grejdi-mbta** [comment]
@paulswartz @thecristen Getting back on this. I renamed this action to be more specific to just Fargate tasks. For Fargate tasks, the network configuration is **required**, but overrides is not. For overrides, I've added a check that runs the command with or without depending on the environment variable being available.

**Review by thecristen — APPROVED** [review]

**Review by paulswartz — APPROVED** [review]
🍰 

You'll need to tag a new version, as well as re-tagging the `v1` tag.

**Comment by grejdi-mbta** [comment]
@paulswartz Tagged, and updated `v1`. Thanks.

**Pull request merged**